simulator/modules/postgres.py

```diff
@@ -218,7 +218,7 @@ def _single_insert_mode(events, use_multiple_tables, batch_size, batch_mode):
     "temperature-min-max": "SELECT max(temperature), min(temperature) FROM events",
     "temperature-stats": "SELECT max(temperature), avg(temperature), min(temperature) FROM events",
     "temperature-stats-per-device": "SELECT device_id, max(temperature), avg(temperature), min(temperature) FROM events GROUP BY device_id",
-    "newest-per-device": "SELECT e.device_id, e.temperature FROM events e JOIN (SELECT device_id, max(timestamp) as ts FROM events GROUP BY device_id) newest ON e.device_id=newest.device_id AND e.timestamp = newest.ts",
+    "newest-per-device": "SELECT device_id, temperature from (SELECT device_id, temperature, timestamp=max(timestamp) over (partition by device_id) newest FROM events) e where newest",
```

Comment:
We do not need more than one row per device, so I would rather suggest `row_number() over (partition by device_id order by timestamp desc) rn .... where rn=1`

Reply:
Why to a sort when all we need is the max? Optimizers may find better access path for the max

In PostgreSQL, a no-sort is possible with your query but this requires a DESC order in the index:
https://dbfiddle.uk/?rdbms=postgres_14&fiddle=bd0e5b5a18969cd9694eb8a8db98ad0c

With my query both index ways are ok